### PR TITLE
new python/mxnet/lr_scheduler.py

### LINKED ISSUE #625: improve the sematic of lr_scheduler
the current lr_scheduler is based on epoch, it is not flexible, and it has problem with the distributed training, where the optimizer runs on the server nodes. a server node neither can get the current epoch number, nor there is a clear definition of epoch for async sgd. 

i suggest to turn to a `batch` based lr scheduler implemented in python/mxnet, while provide an epoch like interface to users, say

```
new_lr = .8 * old_lr for every .5 epoch
```

we then set a correct batch value for lr_scheduler.

but for a correct implementation, `batch` is not a good term. it means reading a data batch, while not necessary a weight updating:

1. we can sum the gradient for several batches and then apply one weight updating. often useful if the model is huge, while gpu mem is little, especially on the distributed setting.
2. on distributed training, assume n machines and b batches, there are n*b weight updatings for async_sgd while only b updatings for sync_sgd

so i suggest to pass `num_updates` to the lr_scheduler. but there there are still two issues

1. we may update layer 1 (conv layer) for each batch, while layer 2 (big fullc layer) for every 2 batches.
2. assume there are two workers, on async_sgd, the possible updating order on the server node could be
```
layer_1_worker_1, layer_2_worker_2, layer_1_worker_2, layer_2_worker_1
``` 
to simplify the thing, we pass "the maximal number of updates applied to a weight" to the lr_scheduler, for the above example, the values are

```
1, 1, 2, 2
```

### PR DESCRIPTION
**Pull request by mli**
this pr resolves #625 

it adds python/mxnet/lr_scheduler.py to replace misc.py, and updates model.py and optimizer.py

the results:

do factor for each epoch
```
~/work/mxnet/example/image-classification $ python train_mnist.py --lr-factor .8 
2015-11-18 22:42:29,104 Node[0] start with arguments Namespace(batch_size=128, data_dir='mnist/', gpus=None, kv_type='local', load_epoch=None, lr=0.1, lr_factor=0.8, lr_factor_epoch=1, model_prefix=None, network='mlp', num_epochs=10, num_examples=60000)
[22:42:30] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(128,784)
[22:42:30] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(128,784)
2015-11-18 22:42:30,578 Node[0] Start training with [cpu(0)]
2015-11-18 22:42:32,657 Node[0] Epoch[0] Train-accuracy=0.692491
2015-11-18 22:42:32,657 Node[0] Epoch[0] Time cost=2.077
2015-11-18 22:42:32,826 Node[0] Epoch[0] Validation-accuracy=0.945212
2015-11-18 22:42:32,827 Node[0] Update[469]: Change learning rate to 0.08000
2015-11-18 22:42:34,841 Node[0] Epoch[1] Train-accuracy=0.958400
2015-11-18 22:42:34,841 Node[0] Epoch[1] Time cost=2.015
2015-11-18 22:42:35,001 Node[0] Epoch[1] Validation-accuracy=0.962841
2015-11-18 22:42:35,002 Node[0] Update[937]: Change learning rate to 0.06400
2015-11-18 22:42:37,085 Node[0] Epoch[2] Train-accuracy=0.975377
2015-11-18 22:42:37,085 Node[0] Epoch[2] Time cost=2.084
2015-11-18 22:42:37,299 Node[0] Epoch[2] Validation-accuracy=0.971354
2015-11-18 22:42:37,300 Node[0] Update[1405]: Change learning rate to 0.05120
```

do factor for each .5 epoch
````
~/work/mxnet/example/image-classification $ python train_mnist.py --lr-factor .8 --lr-factor-epoch .5
2015-11-18 22:44:13,416 Node[0] start with arguments Namespace(batch_size=128, data_dir='mnist/', gpus=None, kv_type='local', load_epoch=None, lr=0.1, lr_factor=0.8, lr_factor_epoch=0.5, model_prefix=None, network='mlp', num_epochs=10, num_examples=60000)
[22:44:14] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(128,784)
[22:44:14] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(128,784)
2015-11-18 22:44:14,890 Node[0] Start training with [cpu(0)]
2015-11-18 22:44:15,939 Node[0] Update[235]: Change learning rate to 0.08000
2015-11-18 22:44:16,925 Node[0] Epoch[0] Train-accuracy=0.693309
2015-11-18 22:44:16,925 Node[0] Epoch[0] Time cost=2.034
2015-11-18 22:44:17,099 Node[0] Epoch[0] Validation-accuracy=0.945613
2015-11-18 22:44:17,100 Node[0] Update[469]: Change learning rate to 0.06400
2015-11-18 22:44:18,097 Node[0] Update[703]: Change learning rate to 0.05120
2015-11-18 22:44:19,066 Node[0] Epoch[1] Train-accuracy=0.959535
2015-11-18 22:44:19,066 Node[0] Epoch[1] Time cost=1.967
2015-11-18 22:44:19,239 Node[0] Epoch[1] Validation-accuracy=0.962540
2015-11-18 22:44:19,240 Node[0] Update[937]: Change learning rate to 0.04096
2015-11-18 22:44:20,212 Node[0] Update[1171]: Change learning rate to 0.03277
2015-11-18 22:44:21,244 Node[0] Epoch[2] Train-accuracy=0.975344
2015-11-18 22:44:21,244 Node[0] Epoch[2] Time cost=2.005
2015-11-18 22:44:21,414 Node[0] Epoch[2] Validation-accuracy=0.973658
2015-11-18 22:44:21,415 Node[0] Update[1405]: Change learning rate to 0.02621
2015-11-18 22:44:22,404 Node[0] Update[1639]: Change learning rate to 0.02097
2015-11-18 22:44:23,389 Node[0] Epoch[3] Train-accuracy=0.981704
2015-11-18 22:44:23,390 Node[0] Epoch[3] Time cost=1.975
2015-11-18 22:44:23,579 Node[0] Epoch[3] Validation-accuracy=0.973257
2015-11-18 22:44:23,580 Node[0] Update[1873]: Change learning rate to 0.01678
2015-11-18 22:44:24,572 Node[0] Update[2107]: Change learning rate to 0.01342
2015-11-18 22:44:25,536 Node[0] Epoch[4] Train-accuracy=0.985710
2015-11-18 22:44:25,536 Node[0] Epoch[4] Time cost=1.957
2015-11-18 22:44:25,694 Node[0] Epoch[4] Validation-accuracy=0.974960
2015-11-18 22:44:25,694 Node[0] Update[2341]: Change learning rate to 0.01074
```

works for distributed sgd
```
~/work/mxnet/example/image-classification $ ../../tracker/dmlc_local.py -n 1 -s 2 python train_mnist.py --kv-type dist_sync --lr-factor .8 
2015-11-18 22:43:15,535 Node[0] start with arguments Namespace(batch_size=128, data_dir='mnist/', gpus=None, kv_type='dist_sync', load_epoch=None, lr=0.1, lr_factor=0.8, lr_factor_epoch=1, model_prefix=None, network='mlp', num_epochs=10, num_examples=60000)
[22:43:16] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(128,784)
[22:43:16] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(128,784)
2015-11-18 22:43:17,004 Node[0] Start training with [cpu(0)]
2015-11-18 22:43:19,824 Node[0] Epoch[0] Train-accuracy=0.692491
2015-11-18 22:43:19,824 Node[0] Epoch[0] Time cost=2.725
2015-11-18 22:43:19,997 Node[0] Epoch[0] Validation-accuracy=0.945212
2015-11-18 22:43:20,002 Server[1] Update[469]: Change learning rate to 0.08000
2015-11-18 22:43:20,002 Server[0] Update[469]: Change learning rate to 0.08000
2015-11-18 22:43:22,652 Node[0] Epoch[1] Train-accuracy=0.958250
2015-11-18 22:43:22,652 Node[0] Epoch[1] Time cost=2.655
2015-11-18 22:43:22,815 Node[0] Epoch[1] Validation-accuracy=0.964844
2015-11-18 22:43:22,820 Server[1] Update[937]: Change learning rate to 0.06400
2015-11-18 22:43:22,820 Server[0] Update[937]: Change learning rate to 0.06400
2015-11-18 22:43:25,496 Node[0] Epoch[2] Train-accuracy=0.975511
2015-11-18 22:43:25,496 Node[0] Epoch[2] Time cost=2.681
2015-11-18 22:43:25,685 Node[0] Epoch[2] Validation-accuracy=0.973658
2015-11-18 22:43:25,690 Server[1] Update[1405]: Change learning rate to 0.05120
2015-11-18 22:43:25,690 Server[0] Update[1405]: Change learning rate to 0.05120
2015-11-18 22:43:28,480 Node[0] Epoch[3] Train-accuracy=0.982188
2015-11-18 22:43:28,480 Node[0] Epoch[3] Time cost=2.795
2015-11-18 22:43:28,710 Node[0] Epoch[3] Validation-accuracy=0.977865
2015-11-18 22:43:28,715 Server[1] Update[1873]: Change learning rate to 0.04096
2015-11-18 22:43:28,715 Server[0] Update[1873]: Change learning rate to 0.04096
```

works for distributed asgd
```
~/work/mxnet/example/image-classification $ ../../tracker/dmlc_local.py -n 1 -s 2 python train_mnist.py --kv-type dist_async --lr-factor .8 
2015-11-18 22:43:42,538 Node[0] start with arguments Namespace(batch_size=128, data_dir='mnist/', gpus=None, kv_type='dist_async', load_epoch=None, lr=0.1, lr_factor=0.8, lr_factor_epoch=1, model_prefix=None, network='mlp', num_epochs=10, num_examples=60000)
[22:43:43] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(128,784)
[22:43:44] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(128,784)
2015-11-18 22:43:44,033 Node[0] Start training with [cpu(0)]
2015-11-18 22:43:46,764 Node[0] Epoch[0] Train-accuracy=0.692491
2015-11-18 22:43:46,768 Node[0] Epoch[0] Time cost=2.642
2015-11-18 22:43:46,925 Node[0] Epoch[0] Validation-accuracy=0.945212
2015-11-18 22:43:46,930 Server[1] Update[469]: Change learning rate to 0.08000
2015-11-18 22:43:46,930 Server[0] Update[469]: Change learning rate to 0.08000
2015-11-18 22:43:49,513 Node[0] Epoch[1] Train-accuracy=0.958250
2015-11-18 22:43:49,513 Node[0] Epoch[1] Time cost=2.588
2015-11-18 22:43:49,675 Node[0] Epoch[1] Validation-accuracy=0.964844
2015-11-18 22:43:49,680 Server[1] Update[937]: Change learning rate to 0.06400
2015-11-18 22:43:49,680 Server[0] Update[937]: Change learning rate to 0.06400
2015-11-18 22:43:52,293 Node[0] Epoch[2] Train-accuracy=0.975511
2015-11-18 22:43:52,293 Node[0] Epoch[2] Time cost=2.618
2015-11-18 22:43:52,483 Node[0] Epoch[2] Validation-accuracy=0.973658
2015-11-18 22:43:52,491 Server[0] Update[1405]: Change learning rate to 0.05120
2015-11-18 22:43:52,491 Server[1] Update[1405]: Change learning rate to 0.05120
2015-11-18 22:43:55,083 Node[0] Epoch[3] Train-accuracy=0.982188
2015-11-18 22:43:55,083 Node[0] Epoch[3] Time cost=2.600
2015-11-18 22:43:55,296 Node[0] Epoch[3] Validation-accuracy=0.977865
2015-11-18 22:43:55,303 Server[1] Update[1873]: Change learning rate to 0.04096
```

### CODE REVIEW

**Comment by jermainewang** [comment]
So now `lr_scheduler` is component of server or client?

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -80,6 +71,20 @@ def set_lr_scale(self, args_lrscale):
...
+        update num_update
+
+        Parameters:
+        index : int
+            The index will be updated
+        """
+        if index not in self._index_update_count:
+            self._index_update_count[index] = 0
+        self._index_update_count[index] += 1
+        self.num_update = max(self._index_update_count[index], self.num_update)
```
Comment (jermainewang) [inline python/mxnet/optimizer.py]: I am not sure about this `max`. In ASGD settings, it is just really weird to let each worker decrease learning rate separately. So it is possible that, some workers are using lr = 0.5 while some workers are using lr = 0.005. What does this mean in fact?

I think a (slightly) better semantic is to tag the weight with version and the learning rate strategy is version-based. Say if the version is larger than some step, decrease the learning rate to some extent.

Comment (mli) [inline python/mxnet/optimizer.py]: both the optimizer and the lr_scheduler run on the server nodes for distributed training. the only inconsistency happens on the asgd setting that different servers may have different update_count, but the difference is upper bounded.

Comment (jermainewang) [inline python/mxnet/optimizer.py]: Ok, i think your `num_updates` are the same as my `version` to some extent. I just feel like having different learning rates per workers are strange. BTW, how momentum is done in ASGD settings right now? Does each worker keep their own momentum of weights?

Comment (mli) [inline python/mxnet/optimizer.py]: all weight updating and the associated data, such as lr and momentum, are on the server nodes. a worker only read data, pull weight, compute gradient, and push gradient, nothing more. so a worker dose not have a learning rate on the distributed training.

**Comment by mli** [comment]
@antinucleon please have a look at this PR, i merged all image examples into `image-classification`, (will handle `distributed-training` later) and also updated `lr-scheduler` (will remove mxnet/misc.py later)

i'll merge this PR if everything goes fine

File: python/mxnet/optimizer.py (hunk 1 above)
Comment (jermainewang) [inline python/mxnet/optimizer.py]: Got it. I'm fine with the change.

**Comment by antinucleon** [comment]
LGTM
